Hot-Rod-Transactions-NonXA.asciidoc

```diff
@@ -0,0 +1,311 @@
+It main goal is to register itself and it will map a transaction to the remote caches involved.
+
+If multiple remote caches are involved in a single transaction, it commit will happen orderly between the remote caches.
+This avoid possible deadlock among the remote caches.
+
+== Transaction's Isolation Level
+
+As said, only `REPEATABLE_READ` will be supported. The reasons are:
+
+* It is the strongest guarantee supported by the server,
+
+* It offers a better performance by skipping interaction with the server at each requests,
+
+* In the future, `READ_COMMITTED` will be removed.
+
+== Description
+
+Each remote-cache will have a transaction context associated to a specific transaction.
+Each key read or write will be stored in this context.
+Also, metadata needed for transaction validation is stored in the context.
+
+At the first operation, the remote cache registers itself as a participant in the `SynchronizationAdapter`.
+
+
+
+=== Transaction execution
+
+Read operations is executed as following:
+
+* If the key exists in transaction context, return its value;
+* Otherwise, fetch the value and metadata from the server and store it in the transaction context.
+
+Write operations is executed as following:
+
+* If the key doesn't exists in the transaction context and it is required (conditional value or needs to return previous value), fetches value and metadata from the server and store it in the transaction context;
+* Applies the operation based on the value in the transaction context and updates its value.
+
+NOTE: The value and metadata can be fetch from any node in the cluster.
+
+=== Transaction commit
+
+A `Transaction` is committed in 2 phases: `beforeCompletion()` and `afterCompletion()`.
+
+In `beforeCompletion()`, the client collects the written keys and sends them to the server (any member in cluster).
+The following information is sent:
+
+* XID - this is the transaction id;
+* For each key, it sends:
+** The key;
+** The `ReadState`;
+** The version read (if any);
+** The new value;
+** The metadata;
+
+[NOTE]
+======
+The `ReadState` represents the following information:
+
+* `NOT_READ`: the key wasn't read before the write;
+* `NON_EXISTING`: the key was read before write and it didn't exist;
+* `READ`: The key was read before write and it exists (the versionr read is required)
+======
+
+In the server side, when the request is received, it replays the transaction.
+A new transaction is created and set its XID based on the XID received (to avoid any XID translation).
+
+For each key received, the server performs the following:
+
+* It checks the `ReadState` value:
+** `NOT_READ`: it doesn't fetch the previous value.
+** `NON_EXISTING`: it checks if the key doesn't exist.
+If the key exists, the transaction is rolled-back.
+** `READ`: it fetches the value and metadata.
+Then it compares the version received with the version stored in the metadata.
+If they don't match, the transaction is rolled-back.
```

Comment:
> Will the client get WriteSkewException?

No, it will only get the XA return/error codes. It is the only information it needs to reply to the TransactionManager.


> So the contacted server (originator) will need to fetch the data from remote node? Wouldn't it be better to prepare the data on each owner from client?

I'm not ready for this kind of optimization. you will end up in `N` local transaction in the server all with the same XID. I think it is too complex for the initial implementation.